apps/condo/domains/common/components/MultipleFileUpload.tsx

```diff
@@ -0,0 +1,201 @@
+    useCreate: (attrs, onComplete) => (attrs) => Promise<DBFile>
+    useUpdate: (attrs, onComplete) => (update, attrs) => Promise<DBFile>
+    useSoftDelete: (attrs, onComplete) => (state, attrs) => Promise<unknown>
+}
+
+const reducer = (state, action) => {
+    const { type, payload: file } = action
+    switch (type) {
+        case 'delete':
+            return {
+                ...state,                    
+                added: [...state.added].filter(addFile => addFile.id !== file.id),
+                deleted: [...state.deleted, file],
+            }
+        case 'add':
+            return {
+                ...state,
+                added: [...state.added, file],
+            }
+        default:
+            throw new Error(`unknown action ${type}`)
+    }
+}
+
+const convertFilesToUploadFormat = (files: DBFile[]): UploadListFile[] => {
+    if (isEmpty(files)) {
+        return []
+    }
+    const uploadFiles = files.map(({ id, file }) => {
+        const fileInList = {
+            uid: file.id,
+            id,
+            name: file.originalFilename,
+            status: 'done' as UploadFileStatus,
+            url: file.publicUrl,
+        }
+        return fileInList        
+    })
+    return uploadFiles
+}
+
+interface IUploadComponentProps {
+    fileList: DBFile[]    
+}
+interface MultipleFileUploadHookArgs {
+    Model: Module
+    relationField: string
+    initialCreateValues?: Record<string, unknown>
+}
+
+export const useMultipleFileUploadHook = ({ 
+    Model, 
+    relationField,
+    initialCreateValues = {},    
+}: MultipleFileUploadHookArgs): [React.FC<IUploadComponentProps>, (id: string) => Promise<void>] => {    
+    const [modifiedFiles, dispatch] = useReducer(reducer, { added: [], deleted: [] })
+    const modifiedFilesRef = useRef(modifiedFiles)
+    // Todo(zuch): without ref modifiedFiles dissappears on submit
+    useEffect(() => {
+        modifiedFilesRef.current = modifiedFiles
+    }, [modifiedFiles])
+
+    const updateAction = Model.useUpdate({}, () => Promise.resolve())
+    const deleteAction = Model.useSoftDelete({}, () => Promise.resolve())
+    
+    const syncModifiedFiles = async (id: string) => {
+        const { added, deleted } = modifiedFilesRef.current
+        for (const file of added) {
+            await updateAction({ [relationField]: id }, file)
+        }
+        for (const file of deleted) {
+            await deleteAction({}, { id: file.id })
+        }
+    }
+
+    const UploadComponent: React.FC<IUploadComponentProps> = useMemo(() => {
+        const UploadWrapper = ({ fileList }) => (
+            <MultipleFileUpload
+                fileList={fileList}
+                initialCreateValues={{ ...initialCreateValues, [relationField]: null }}
+                Model={Model}
+                onFilesChange={dispatch}
+            />
+        )
+        return UploadWrapper
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [])
+    return [
```

Comment:
It's more common to return an object instead of an array if you plan to export something else in the future.

Reply:
Done